src/plugins/crypto/crypto.c

```diff
@@ -0,0 +1,220 @@
+#endif
+#include "crypto.h"
+#include "gcrypt_operations.h"
+
+
+/**
+ * @brief open the crypto plugin
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoOpen(Plugin *handle ELEKTRA_UNUSED, Key *errorKey)
+{
+	return elektraCryptoInit(errorKey);
+}
+
+/**
+ * @brief close the crypto plugin
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoClose(Plugin *handle ELEKTRA_UNUSED, Key *errorKey ELEKTRA_UNUSED)
+{
+	return 1;
+}
+
+/**
+ * @brief establish the Elektra plugin contract and decrypt values, if possible.
+ *
+ * The crypto configuration is expected to be contained within the KeySet ks.
+ * All keys having a metakey "crypto/encrypted" with a strlen() > 0 are being decrypted.
+ *
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoGet(Plugin *handle ELEKTRA_UNUSED, KeySet *ks, Key *parentKey)
+{
+	Key *k;
+	elektraCryptoHandle *cryptoHandle;
+
+	// Publish module configuration to Elektra (establish the contract)
+	if (!strcmp (keyName(parentKey), "system/elektra/modules/crypto"))
+	{
+		KeySet *moduleConfig = ksNew (30,
+#include "contract.h"
+			KS_END);
+		ksAppend (ks, moduleConfig);
+		ksDel (moduleConfig);
+		return 1;
+	}
+
+	// the actual decryption
+
+	// for now we expect the crypto configuration to be stored in the KeySet ks
+	// we may add more options in the future
+
+	if(elektraCryptoHandleCreate(&cryptoHandle, ks, parentKey) != 1)
+	{
+		goto error;
+	}
+
+	ksRewind (ks);
+	while ((k = ksNext (ks)) != 0)
+	{
+		if(elektraCryptoDecrypt(cryptoHandle, k, parentKey) != 1)
+		{
+			goto error;
+		}
+	}
+	elektraCryptoHandleDestroy(cryptoHandle);
+	return 1;
+
+error:
+	elektraCryptoHandleDestroy(cryptoHandle);
+	return -1;
+}
+
+/**
+ * @brief Encrypt values marked for encryption.
+ *
+ * If a key has the metakey "crypto/encrypt" with a strlen() > 0, then the value
+ * will be encrypted using the configuration stored in the KeySet ks.
+ *
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoSet(Plugin *handle ELEKTRA_UNUSED, KeySet *ks, Key *parentKey)
+{
+	Key *k;
+	elektraCryptoHandle *cryptoHandle;
+
+	// for now we expect the crypto configuration to be stored in the KeySet ks
+	// we may add more options in the future
+
+	if(elektraCryptoHandleCreate(&cryptoHandle, ks, parentKey) != 1)
+	{
+		goto error;
+	}
+
+	ksRewind (ks);
+	while ((k = ksNext (ks)) != 0)
+	{
+		if(elektraCryptoEncrypt(cryptoHandle, k, parentKey) != 1)
+		{
+			goto error;
+		}
+	}
+	elektraCryptoHandleDestroy(cryptoHandle);
+	return 1;
+
+error:
+	elektraCryptoHandleDestroy(cryptoHandle);
+	return -1;
+}
+
+/**
+ * @brief TBD
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoError(Plugin *handle ELEKTRA_UNUSED, KeySet *ks ELEKTRA_UNUSED, Key *parentKey ELEKTRA_UNUSED)
+{
+	return 1;
+}
+
+/**
+ * @brief initialize the crypto backend
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoInit(Key *errorKey)
```

Comment:
can you give me an example, please? is there already a plugin that uses these functions with internal parameters?